spec: add CLI notation certificate and key specs (#361)
notaryproject/notation: specs/commandline/key.md

# notation key

## Description

> Use ```notation key``` command to manage keys used for signing. User can add/update/list/remove key to/from signing key list. Please be noted this command doesn't manage the lifecycle of signing key itself, it manages the signing key list.

Reviewer: it manages the signing key list only.

## Outline

### notation key command

```text
Manage keys used for signing

Usage:
  notation key [command]

Available Commands:
  add         Add key to signing key list
  list        List keys used for signing
  remove      Remove key from signing key list
  update      Update key in signing key list

Flags:
  -h, --help   help for key
```

### notation key add

```text
Add key to signing key list

Usage:
  notation key add <key_name> [key_path cert_path] [flags]

Flags:
  -d, --default               mark as default
  -h, --help                  help for add
      --id string             key id (required if --plugin is set)
  -p, --plugin string         signing plugin name

>   -c, --plugin-config string   list of comma-separated {key}={value} pairs that are passed as is to the plugin, refer plugin documentation to set appropriate values

Reviewer 2: The indentation should be fixed this line.

>       --plugin-config string  list of comma-separated {key}={value} pairs that are passed as is to the plugin, refer plugin documentation to set appropriate values

Reviewer 3: It is better to not have `comma-separated {key}={value} pairs` since `value` itself MAY have `,`.
```suggestion
      --plugin-config strings {key}={value} pairs that are passed as is to the plugin, refer plugin documentation to set appropriate values
```

For example,
```shell
notation key add -p test --plugin-config foo=fun --plugin-config bar=beer test_key
```

```

### notation key list

```text
List keys used for signing

Usage:

>   notation key list [flags]

Reviewer 4: what kind of flags are we thinking here?

Reviewer 3: It seems currently we only have `--help`.

Aliases:
  list, ls

Flags:
  -h, --help   help for list

```

> ### notation key remove

Reviewer 3: `remove` is not consistent with `notation cert delete`. We should make the command names consistent.

Usage:
  notation key remove <key_name>... [flags]

Aliases:
  remove, rm

Flags:
  -h, --help   help for remove
```

### notation key update

```text
Update key in signing key list

Usage:
  notation key update <key_name> [flags]

Aliases:
  update, set

Flags:
  -d, --default   mark as default
  -h, --help      help for update

```

## Usage

### Add a local key to signing key list and make it as default signing key

```text
notation key add <key_name> <key_file_path> <cert_file_path> --default
```

Upon successful adding, a key name is printed out for added signing key with additional info "marked as default".

### Add a default signing key referencing the key identifier for the remote key, and the plugin associated with it

```text
notation key add <key_name> --plugin <plugin_name> --id <remote_key_id> --default
```

Upon successful adding, a key name is printed out for added signing key with additional info "marked as default".

### Update the default signing key

```text
notation key update <key_name> --default
```

Upon successful update, the supplied key name is printed out with additional info "marked as default".

### List signing keys

```text
notation key list
```

> Upon successful execution, a list of keys is printed out with information of name, key path, certificate path, key id and plugin name. The default signing key name is preceded by an asterisk. The key id and plugin name are used together to provide the information of the key identifier for the remote key, and the plugin associated with it.

Reviewer 2: ```suggestion
Upon successful execution, a list of keys are printed out with information of name, key path, certificate path, key id and plugin name. The default signing key name is preceded by an asterisk. The key id and plugin name are used together to provide the information of the key identifier for the remote key, and the plugin associated with it.
```

> Upon successful execution, a list of keys is printed out with information of name, key path, certificate path, key id and plugin name. The default signing key name is preceded by an asterisk. The key id and plugin name are used together to provide the information of the key identifier for the remote key, and the plugin associated with it.

Reviewer: ```suggestion
Upon successful execution, a list of keys is printed out with information of name, key path, certificate path, key id and plugin name. The default signing key name is preceded by an asterisk. The key id and plugin name are used together to provide the information of the key identifier for the remote key and the plugin associated with it.
```

### Remove sign keys

```text

> notation key remove <key_name_1> <key_name_2>

Reviewer 2: notation key remove <key_name>...

Reviewer 2: what's the difference between this command and `notation key remove <key_name>... [flags]` on line 65

Author: The first remove is the command line spec. The second one is an example of removing two keys. I will update the description of usage to make it clear.

```

> Upon successful execution, the names of removed signing keys are printed out. Please be noted if default signing key is removed, there is no default signing key assigned automatically. User need to update the default signing key explicitly.

Reviewer: ```suggestion
Upon successful execution, the names of removed signing keys are printed out. Please be noted if default signing key is removed, Notation will not automatically assign a new default signing key. User needs to update the default signing key explicitly.
```

# notation key

## Description

Use ```notation key``` command to manage keys used for signing. User can add/update/list/remove key to/from signing key list. Please be noted this command doesn't manage the lifecycle of signing key itself, it manages the signing key list only.

## Outline

### notation key command

```text
Manage keys used for signing

Usage:
  notation key [command]

Available Commands:
  add         Add key to signing key list
  list        List keys used for signing
  delete      Delete key from signing key list
  update      Update key in signing key list

Flags:
  -h, --help  help for key
```

### notation key add

```text
Add key to signing key list

Usage:

>   notation key add [flags] <key_name> [key_path cert_path]

Reviewer 4: By exposing this command with `[key_path cert_path]` we are saying, notation support location with unprotected keys only.

If we are saying we dont support signing with local keys we should remove [`key_path cert_path]`. If we are supporting local signing we should add support for accepting password.

Flags:
  -d, --default                mark as default
  -h, --help                   help for add
      --id string              key id (required if --plugin is set)
  -p, --plugin string          signing plugin name
      --plugin-config strings  {key}={value} pairs that are passed as is to the plugin, refer plugin documentation to set appropriate values
```

### notation key list

```text
List keys used for signing

Usage:
  notation key [flags] list

Aliases:
  list, ls

Flags:
  -h, --help                   help for list

```

### notation key delete

```text
Delete key from signing key list

Usage:
  notation key delete [flags] <key_name>...

Aliases:
  delete, rm

Flags:
  -h, --help                   help for delete

```

### notation key update

```text
Update key in signing key list

Usage:

>   notation key update [flags] <key_name>

Reviewer 4: Whats the purpose/ intended use of update command? Is it just to update default key?

Author: Currently only one flag '--default', there could be more later.
